prov/lpp/src/lpp_hmem.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2019-2023 GigaIO, Inc. All Rights Reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "lpp.h"
+#include "ofi_hmem.h"
+
+#if HAVE_CUDA || HAVE_ROCR
+
+/* lpp_hmem_ptr() - Checks if a pointer is associated with an hmem_iface
+ * @lpp_domain:	The lpp_domain that wants to use the memory
+ * @ptr:	The memory in question
+ * @iface:	Returns the fi_hmem_iface the ptr belongs to
+ *
+ *		If no hmem_ifaces are enabled, then FI_HMEM_SYSTEM is returned
+ *		Only 1 interface can be configured at a time, i.e. systems with
+ *		a mix of Nvidia and AMD gpus won't work.
+ */
+int lpp_hmem_ptr(Lpp_domain_t *lpp_domain, const void *ptr,
+				  enum fi_hmem_iface *iface)
+{
+	if (!iface)
+		return -FI_EINVAL;
+
+	*iface = FI_HMEM_SYSTEM;
+
+	if (lpp_domain->enabled_iface == FI_HMEM_SYSTEM)
+		return 0;
+
+	if (lpp_domain->enabled_iface == FI_HMEM_ROCR
+	   && rocr_is_addr_valid(ptr, NULL, NULL)) {
+		*iface = FI_HMEM_ROCR;
+	} else if (lpp_domain->enabled_iface == FI_HMEM_CUDA
+		&& cuda_is_addr_valid(ptr, NULL, NULL)) {
+
+		#if HAVE_CUDA
+		//TODO: not entirely clear if this is required for write-only MRs.
+		// Might wan to unset later as this can affect performance.
+		CUresult cures;
+		int flag = 1;
+		cures = ofi_cuPointerSetAttribute(&flag,
```

Comment:
We have problem with this sync memops set when running with NCCL >= 2.19. Basically it's not safe to set `CU_POINTER_ATTRIBUTE_SYNC_MEMOPS` for NCCL application

Reply:
I guess you set this attribute to enforce a synchronous cuda memcpy inside libfabric, while cuda memcpy is not allowed inside Libfabric either for NCCL application - https://github.com/ofiwg/libfabric/issues/6124

Reply:
Our device uses GPUDirectRDMA, which avoid the need for cudaMemcpy. We're setting the SYNC_MEMOPS attr out of an abundance of caution, as outlined in GpuDirectRDMA docs (https://docs.nvidia.com/cuda/gpudirect-rdma/#synchronization-and-memory-ordering).

Poking through the code, the EFA rdm calls `efa_rdm_attempt_to_sycm_memops` before a lot of transfers. How are you guys handing this issue then? Is there some nuance to when 'efa_mr->needs_sync` is set?

As for NCCL >=2.19, is this related to the changes with device memory allocation? I ran into a problem testing out newer NCCL version, but was able to mitigate it with the `NCCL_CUMEM_ENABLE=0` env var. 

Also, now that I'm noticing it, this could probably be reworked to use `cuda_set_sync_memops`...

Reply:
>As for NCCL >=2.19, is this related to the changes with device memory allocation? I ran into a problem testing out newer NCCL version, but was able to mitigate it with the NCCL_CUMEM_ENABLE=0 env var.

Yes, the problem can be mitigated by that env exactly. We have a boolean as `FI_OPT_CUDA_API_PERMITTED` to toggle whether CUDA API can be permitted to call inside Libfabric in data transfer. NCCL plugin set this boolean to false to disable cuda api usage in Libfabric.


https://ofiwg.github.io/libfabric/main/man/fi_endpoint.3.html

```

FI_OPT_CUDA_API_PERMITTED - bool
    This option only applies to the fi_setopt call. It is used to control endpoint’s behavior in making calls to CUDA API. By default, an endpoint is permitted to call CUDA API. If user wish to prohibit an endpoint from making such calls, user can achieve that by set this option to false. If an endpoint’s support of CUDA memory relies on making calls to CUDA API, it will return -FI_EOPNOTSUPP for the call to fi_setopt. If either CUDA library or CUDA device is not available, endpoint will return -FI_EINVAL. All providers that support FI_HMEM capability implement this option.


```

The problem is that boolean is set in the EP level via fi_setopt. while the MR reg can happen before that, so we cannot reuse that boolean so we can only do the toggle in the transmission call.